src/IntegralArrays.jl

```diff
@@ -1,5 +1,73 @@
 module IntegralArrays
 
 # package code goes here
+using Images, IntervalSets
+
+import Base: size, getindex, linearindexing
+
+export IntegralArray
+
+"""
+```
+integral_array = IntegralArray(A)
+```
+
+Returns the integral array of an array. The integral array is calculated by assigning
+to each cell the sum of all cells above it and to its left, i.e. the rectangle from
+(1, 1) to the pixel. An integral array is a data structure which helps in efficient 
+calculation of sum of pixels in a rectangular subset of an array. 
+
+```
+sum = integral_array[ytop..ybot, xtop..xbot]
+sum = integral_array[y ± σy, x ± σx]
+sum = integral_array[CartesianIndex(y, x) ± σ]
+```
+
+The sum of a window in an array can be directly calculated using four array references of 
+the integral array, irrespective of the size of the window, given the `yrange` and `xrange` 
+of the window. Given an integral array - 
+        
+        A - - - - - - B -
+        - * * * * * * * -
+        - * * * * * * * -
+        - * * * * * * * -
+        - * * * * * * * -
+        - * * * * * * * -
+        C * * * * * * D - 
+        - - - - - - - - -
+
+The sum of pixels in the area denoted by * is given by S = D + A - B - C.
+"""
+immutable IntegralArray{T, N, A} <: AbstractArray{T, N}
+    data::A
+end
+
+function IntegralArray(array::AbstractArray)
+    integral_array = Array{Images.accum(eltype(array))}(size(array))
+    sd = coords_spatial(array)
+    cumsum!(integral_array, array, sd[1])
+    for i = 2:length(sd)
+        cumsum!(integral_array, integral_array, sd[i])
+    end
+    IntegralArray{eltype(array), ndims(array), typeof(array)}(integral_array)
+end
+
+linearindexing(A::IntegralArray) = Base.LinearFast()
+size(A::IntegralArray) = size(A.data)
+getindex(A::IntegralArray, i::Int...) = A.data[i...]
+
+getindex(A::IntegralArray, ids::Tuple...) = getindex(A, ids[1]...)
+
+function getindex(A::IntegralArray, i::ClosedInterval...)
```

Comment:
Agreed this may not be entirely simple. Of course you should be able to calculate the actual pixel values from the integral image, and from that you should be able to calculate the actual sub-pixel integral image (where you assume the image to be piecewise constant, aka nearest-neighbor interpolation). I haven't sat down and tried to do this calculation myself yet, so I don't have a sense for how messy it will be. But the advantage is that it will automatically be consistent with the on-pixel integral image, which is a good thing.